ndk/src/native_window.rs

```diff
@@ -126,4 +126,61 @@ impl NativeWindow {
     pub unsafe fn to_surface(&self, env: *mut JNIEnv) -> jobject {
         ffi::ANativeWindow_toSurface(env, self.ptr().as_ptr())
     }
+
+    /// Lock the window's next drawing surface for writing.
+    ///
+    /// Optionally pass the region you intend to draw into `dirty_bounds`.  When this function
+    /// returns it is updated (commonly enlarged) with the actual area the caller needs to redraw.
+    pub fn lock(
+        &self,
+        dirty_bounds: Option<&mut ffi::ARect>,
+    ) -> Result<NativeWindowBufferLockGuard> {
+        let dirty_bounds = match dirty_bounds {
+            Some(dirty_bounds) => dirty_bounds,
+            None => std::ptr::null_mut(),
+        };
+        let mut out = unsafe { std::mem::zeroed() };
+        let ret = unsafe { ffi::ANativeWindow_lock(self.ptr.as_ptr(), &mut out, dirty_bounds) };
+        status_to_io_result(ret, NativeWindowBufferLockGuard(out, self))
+    }
+}
+
+/// Lock holding the next drawing surface for writing.  It is unlocked and posted on [`drop()`].
+#[derive(Debug)]
+pub struct NativeWindowBufferLockGuard<'a>(ffi::ANativeWindow_Buffer, &'a NativeWindow);
+
+impl<'a> NativeWindowBufferLockGuard<'a> {
+    /// The number of pixels that are shown horizontally.
+    pub fn width(&self) -> i32 {
+        self.0.width
+    }
+
+    // The number of pixels that are shown vertically.
+    pub fn height(&self) -> i32 {
+        self.0.height
+    }
+
+    /// The number of _pixels_ that a line in the buffer takes in memory.
+    ///
+    /// This may be `>= width`.
+    pub fn stride(&self) -> i32 {
+        self.0.stride
+    }
+
+    /// The format of the buffer. One of [`HardwareBufferFormat`].
+    pub fn format(&self) -> HardwareBufferFormat {
+        unsafe { std::mem::transmute(self.0.format) }
```

Comment:
might get undefined behaviour here if the implementation isn't updated before new formats are potentially exposed on Android.

Could potentially use TryFromPrimitive from the num_enum crate or implement something equivalent manually that will ensure the u32 from the NDK is a known Rust enum variant.

Reply:
oh, yeah, you're already using `num_enum` here and have `TryFromPrimitive`

Reply:
Yeah indeed, that's sloppy. Shall I adopt the `.unwrap()` or return `Result<, TryFromPrimitive>` for both?

Reply:
yeah tricky. we already know there are r26 formats that aren't added here so the possibility of unknown formats seems worth considering.

in practice I'd guess only a small number of common formats ever show up as formats you'd see for rendering to a window from the cpu though - most of these formats probably only really show up in connection with GLES/vulkan.